### ISSUE TITLE
Reprex to debug dplyr linking

### ISSUE THREAD

**Issue by jayhesselberth** [tidyverse/dplyr#2649]
I'm seeing a bunch of errors when I try to link to the new dplyr via `LinkingTo:` including:

- a header guard typo in `dplyr/workarounds/dplyr.h`
- an include of a missing file `tools/DotsOf.h` in `tools/tools.h`

Once those are fixed you hit:

``` cpp
In file included from /usr/local/lib/R/3.3/site-library/dplyr/include/dplyr.h:8:
In file included from /usr/local/lib/R/3.3/site-library/dplyr/include/dplyr/dplyr.h:4:
In file included from /usr/local/lib/R/3.3/site-library/dplyr/include/dplyr/registration.h:15:
/usr/local/lib/R/3.3/site-library/dplyr/include/dplyr_RcppExports.h:7:10: fatal error: 'dplyr_types.h' file not found
#include "dplyr_types.h"
         ^
1 error generated.
```
 I am not sure why `dplyr_types.h` is there but looks to be auto-generated by `Rcpp::compileAttributes()`. If I go in and manually delete that include in the installed header, I get a clean compilation.

Here's a reprex:

``` r
setwd(tempdir())
devtools::create("foopkg")
setwd("foopkg")

devtools::use_rcpp()

cat(
  "LinkingTo: Rcpp, BH, dplyr, plogr, bindrcpp",
  file = "DESCRIPTION",
  sep = "\n",
  append = TRUE
)

cat(
  c("#' @useDynLib foopkg",
    "NULL"),
  file = "R/foo.R",
  sep = "\n"
)

cat(
  c("// [[Rcpp::export]]",
    "int foo() {return 1;}"),
  file = "src/foo.cpp",
  sep = "\n"
)

devtools::document(roclets=c('rd', 'namespace'))
```

**Comment by jayhesselberth** [comment]
It looks the intent of b51495536c46d9b83561d8670b4188da42730290 was to prevent inclusion of the `dplyr_types.h` header file in RcppExports.cpp. Perhaps something similar could be done to prevent its inclusion in `dplyr_RcppExports.h`?

**Comment by krlmlr** [comment]
I believe we need `dplyr_types.h` in `inst/include`, perhaps with a forwarder in `src` that simply does `#include <dplyr_types.h>`. Would you mind changing that in your PR?

**Comment by jayhesselberth** [comment]
Is `inst/include/dplyr_types.h` just an empty file?

If I add an empty `inst/include/dplyr_types.h` and add `#include <dplyr_types.h>` to `src/dplyr_types.h`, I get clean linking. But I end up with duplicate includes in `dplyr_RcppExports.h`:

``` cpp
// Generated by using Rcpp::compileAttributes() -> do not edit by hand
// Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#ifndef RCPP_dplyr_RCPPEXPORTS_H_GEN_
#define RCPP_dplyr_RCPPEXPORTS_H_GEN_

#include "dplyr_types.h"
#include "dplyr_types.h"
#include <Rcpp.h>
...
```
I suspect that because there are now two `dplyr_types.h` files, both are added?

https://github.com/RcppCore/Rcpp/blob/f99737edbbe699c0128a735518af91a7f1d986da/R/Attributes.R#L432-L448

**Comment by krlmlr** [comment]
Perhaps we can just move `dplyr_types.h` to `inst/include`?

**Comment by jayhesselberth** [comment]
Seems to be fine. Incorporated into #2660.

**Comment by jayhesselberth** [comment]
The above reprex now yields the same error in #2668:

``` cpp
ccache g++ -I/usr/local/Cellar/r/3.3.3_1/R.framework/Resources/include -DNDEBUG  -I/usr/local/opt/gettext/include -I/usr/local/opt/readline/include -I/usr/local/include -I"/usr/local/lib/R/3.3/site-library/Rcpp/include" -I"/usr/local/lib/R/3.3/site-library/BH/include" -I"/usr/local/lib/R/3.3/site-library/dplyr/include" -I"/usr/local/lib/R/3.3/site-library/plogr/include" -I"/usr/local/lib/R/3.3/site-library/bindrcpp/include"   -fPIC  -g -O2  -c RcppExports.cpp -o RcppExports.o
In file included from RcppExports.cpp:4:
In file included from /usr/local/lib/R/3.3/site-library/dplyr/include/dplyr.h:8:
In file included from /usr/local/lib/R/3.3/site-library/dplyr/include/dplyr/dplyr.h:4:
In file included from /usr/local/lib/R/3.3/site-library/dplyr/include/dplyr/registration.h:15:
In file included from /usr/local/lib/R/3.3/site-library/dplyr/include/dplyr_RcppExports.h:7:
In file included from /usr/local/lib/R/3.3/site-library/dplyr/include/dplyr_types.h:5:
/usr/local/lib/R/3.3/site-library/dplyr/include/dplyr/GroupedDataFrame.h:71:15: error: use of undeclared identifier 'build_index_cpp'
      data_ = build_index_cpp(data_);
              ^
1 error generated.
make: *** [RcppExports.o] Error 1
ERROR: compilation failed for package ‘foopkg’
* removing ‘/private/var/folders/_1/4wg2xbj12_dft9p3kq4005n40000gn/T/RtmpKUNzgk/devtools_install_592b1917c73f/foopkg’
Error: Command failed (1)
```

**Comment by jayhesselberth** [comment]
Closing in favor of #2668.
